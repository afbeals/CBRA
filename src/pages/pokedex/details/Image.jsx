// Extenal
import React, { useState } from 'react';
import PropTypes from 'prop-types';

// Local
import { IconButton } from '~Components';
// Constants
const classname = 'image';
/* Declarations */
/* Functions */

// Component
const Image = ({ pClassname, sprites, name }) => {
  const [displayBack, updateDisplayBack] = useState(false);
  const [displayFemale, updateDisplayFemale] = useState(false);

  const maleBack = sprites?.['back_default'] ?? null;
  const femaleBack = sprites?.['back_female'] ?? null;
  const maleFront = sprites?.['front_default'] ?? null;
  const femaleFront = sprites?.['front_female'] ?? null;

  const handleUpdateDisplayBack = () => {
    updateDisplayBack(pVal => !pVal);
  };

  const handleUpdateDisplayFemale = val => {
    updateDisplayFemale(val);
  };

  return (
    <div data-testid="details-image" className={`${pClassname}__${classname}`}>
      <div className={`${pClassname}__${classname}__gender`}>
        <IconButton
          disabled={!maleFront && !maleBack}
          className="male"
          onClick={() => handleUpdateDisplayFemale(false)}
          primary={displayFemale}
        >
          <i className="fas fa-mars" />
        </IconButton>
        <IconButton
          disabled={!femaleFront && !femaleBack}
          className="female"
          onClick={() => handleUpdateDisplayFemale(true)}
          primary={!displayFemale}
        >
          <i className="fas fa-venus" />
        </IconButton>
      </div>
      {sprites && displayFemale && (
        <img
          className={`${pClassname}__${classname}__pokemon`}
          src={displayBack ? femaleBack : femaleFront}
          alt={`${name}`}
        />
      )}
      {sprites && !displayFemale && (
        <img
          className={`${pClassname}__${classname}__pokemon`}
          src={!displayBack ? maleFront : maleBack}
          alt={`${name}`}
        />
      )}
      <div className={`${pClassname}__${classname}__flipper`}>
        <IconButton
          onClick={handleUpdateDisplayBack}
          disabled={!sprites || !!sprites.length}
        >
          <i className="fas fa-sync-alt" />
        </IconButton>
      </div>
    </div>
  );
};

Image.propTypes = {
  pClassname: PropTypes.string.isRequired,
  sprites: PropTypes.object,
  name: PropTypes.string,
};

export default Image;

export { Image };
