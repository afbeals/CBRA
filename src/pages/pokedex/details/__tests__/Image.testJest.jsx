// External
import React from 'react';
import { render, cleanup } from '@testing-library/react';

// Local
import Image from '../Image';

// tests
describe('Image', () => {
  afterEach(cleanup);
  it('should match Image to snapshot', () => {
    const { getByTestId } = render(<Image pClassname={'parent'} />);
    const image = getByTestId('details-image');
    expect(image).toMatchSnapshot();
  });

  it('should have a disabled female icon', () => {
    const sprites = {
      back_default: 'back',
      front_default: 'front',
    };
    const { getAllByRole } = render(
      <Image sprites={sprites} pClassname={'parent'} />,
    );
    const femaleIcon = getAllByRole((content, node) => {
      if (content === 'button') {
        if (node.classList.contains('female')) return true;
      }
      return false;
    })[0];
    expect(femaleIcon.disabled).toBe(true);
  });
});
