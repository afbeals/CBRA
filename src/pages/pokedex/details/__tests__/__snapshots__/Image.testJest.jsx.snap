// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should match Image to snapshot 1`] = `
<div
  class="parent__image"
  data-testid="details-image"
>
  <div
    class="parent__image__gender"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root male iconButton MuiIconButton-colorSecondary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
      data-testid="iconButton"
      disabled=""
      role="button"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <i
          class="fas fa-mars"
        />
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root female iconButton MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
      data-testid="iconButton"
      disabled=""
      role="button"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <i
          class="fas fa-venus"
        />
      </span>
    </button>
  </div>
  <div
    class="parent__image__flipper"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root iconButton MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
      data-testid="iconButton"
      disabled=""
      role="button"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <i
          class="fas fa-sync-alt"
        />
      </span>
    </button>
  </div>
</div>
`;
